### PR TITLE
[feat]: [PL-28305]: adding gcp sm config dto and config update dto

### PR DESCRIPTION
**Pull request by shreyas-harness**
## Describe your changes
Adding google secret manager config dto and config update dto. 
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38098)

File: 930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java [hunk 1]
```diff
@@ -116,6 +117,11 @@ public List<ExecutionCapability> fetchRequiredExecutionCapabilities(ExpressionEv
 
   @Override
   public SecretManagerConfigDTO toDTO(boolean maskSecrets) {
-    return null;
+    GcpSecretManagerConfigDTO configDTO =
+        GcpSecretManagerConfigDTO.builder().delegateSelectors(getDelegateSelectors()).build();
+    if (!maskSecrets) {
+      configDTO.setCredentialsRef(getCredentials());
```
Comment (meenakshi-harness) [inline 930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java:123]: Why we need this check. Credential ref is an actual secret or just a reference to a secret file?

File: 940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigDTO.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Setter
+@SuperBuilder
+@NoArgsConstructor
+@ToString(exclude = {"credentialsRef"})
+@EqualsAndHashCode(callSuper = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Schema(name = "GcpSecretManagerConfig", description = "This contains the information for the Google Secret Manager.")
+public class GcpSecretManagerConfigDTO extends SecretManagerConfigDTO {
+  @Schema(description = SecretManagerDescriptionConstants.GOOGLE_SECRET_MANAGER_CREDENTIALS)
+  private char[] credentialsRef;
```
Comment (meenakshi-harness) [inline 940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigDTO.java:35]: required?

File: 940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigUpdateDTO.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+@OwnedBy(PL)
+@Getter
+@Setter
+@SuperBuilder
+@NoArgsConstructor
+@ToString(exclude = {"credentialsRef"})
+@EqualsAndHashCode(callSuper = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class GcpSecretManagerConfigUpdateDTO extends SecretManagerConfigUpdateDTO {
+  private char[] credentialsRef;
```
Comment (meenakshi-harness) [inline 940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigUpdateDTO.java:32]: same

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

**Review by shreyas-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)* status: 3 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java` line 123 at r1](https://reviewable.io/reviews/harness/harness-core/38098#-ND6alOuDltLeAiKwB6Y-r1-123:-ND7h_Ii2cDFXpZSIeYr:bbagpqi) ([raw file](https://github.com/harness/harness-core/blob/a991b4f5b5a56d17b3dbd67c5efd0da32cb781af/930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java#L123)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Why we need this check. Credential ref is an actual secret or just a reference to a secret file?
</blockquote></details>

It is actual secret. Renamed the variable to credentials.
___
*[`940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigDTO.java` line 35 at r1](https://reviewable.io/reviews/harness/harness-core/38098#-ND6alOuDltLeAiKwB6Z-r1-35:-ND7hYgo-Mc-eRLS2Ndo:b-l3pzke) ([raw file](https://github.com/harness/harness-core/blob/a991b4f5b5a56d17b3dbd67c5efd0da32cb781af/940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigDTO.java#L35)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

required?
</blockquote></details>

Yes, added not empty check
___
*[`940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigUpdateDTO.java` line 32 at r1](https://reviewable.io/reviews/harness/harness-core/38098#-ND6alOuDltLeAiKwB6_-r1-32:-ND7hZQSBMTYuFkaJ57M:b-l3pzke) ([raw file](https://github.com/harness/harness-core/blob/a991b4f5b5a56d17b3dbd67c5efd0da32cb781af/940-secret-manager-client/src/main/java/io/harness/secretmanagerclient/dto/GcpSecretManagerConfigUpdateDTO.java#L32)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same
</blockquote></details>

Yes, added not empty check

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)* status: 3 file(s) to review. 2 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java` line 119 at r3](https://reviewable.io/reviews/harness/harness-core/38098#-ND8HfwL8jvc5OSC9P0J:-ND8HfwL8jvc5OSC9P0K:b-kcd8be) ([raw file](https://github.com/harness/harness-core/blob/5b02f461f60006692eb10e5721e059f9d664841a/930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java#L119)):*
> ```Java
> 
>   @Override
>   public SecretManagerConfigDTO toDTO(boolean maskSecrets) {
> ```

Why this change? As part of framework I can see that separate Mapper class is created because it can involve complex logic.
e.g. AwsSMConfigDTOMapper

**Review by shreyas-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)* status: 3 file(s) to review. 2 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java` line 119 at r3](https://reviewable.io/reviews/harness/harness-core/38098#-ND8HfwL8jvc5OSC9P0J:-ND8UY9b5f-IIO_pRpuQ:bpzieee) ([raw file](https://github.com/harness/harness-core/blob/5b02f461f60006692eb10e5721e059f9d664841a/930-delegate-tasks/src/main/java/software/wings/beans/GcpSecretsManagerConfig.java#L119)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Why this change? As part of framework I can see that separate Mapper class is created because it can involve complex logic.
e.g. AwsSMConfigDTOMapper
</blockquote></details>

AwsSMConfigDTOMapper does not convert AwsSMConfig to AwsSMConfigDTO. 
The point of having toDTO method in SMConfig would be to convert SMConfig to SMConfigDTO.

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)* status: 3 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38098)* status: :shipit: complete! ✓

**Pull request merged**